### PR TITLE
Add formatting script that rejoins line-broken strings and then reapplies clang-format

### PR DESCRIPTION
**Pull request by biddisco**
Rejoin broken strings and then reapply clang-format to improve string layout

### CODE REVIEW

File: tools/string-join.awk [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2019 The STE||AR-Group
```
Comment (msimberg) [inline tools/string-join.awk:1]:
```suggestion
# Copyright (c) 2023 ETH Zurich
```
Maybe also on the formatting script.

File: tools/string-join.awk [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    buffer=buf
+}
+
+BEGIN {
+    enabled=1
+    in_string=0 
+}
+
+# if we see "clang-format off" turn off processing
+# if we see "clang-forma on" turn back on processing
```
Comment (msimberg) [inline tools/string-join.awk:31]:
```suggestion
# if we see "clang-format on" turn back on processing
```

File: tools/string-join.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+# iterate over cpp/etc files and join strings that are broken
+# then, rerun clang-format to rebreak the strings into good sizes
+
+SCRIPT_DIR=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)
+awkscript=$SCRIPT_DIR/string-join.awk
+tmpfile="$(mktemp /tmp/XXXXXXXXX.cpp)" || exit 1
+
+for file in $(git ls-files | grep -E "\.(cpp|hpp|cu)(\.in)?$"); do
+    awk -f $awkscript ${file} >${tmpfile}
```
Comment (msimberg) [inline tools/string-join.sh:17]:
```suggestion
    awk -f "$awkscript" "${file}" >"${tmpfile}"
```
and the same for the other codacy suggestions below (it's all about double-quoting variables).

**Review by msimberg — CHANGES_REQUESTED** [review]
Thanks @biddisco. This is really nice.

Could you please update this to use the new formatting script: https://github.com/pika-org/pika-algorithms/blob/3e4ea7cd7f5eec8eb9c462beae0076c550aa0151/.circleci/config.yml#L104?

**Comment by biddisco** [comment]
> Thanks @biddisco. This is really nice.
> 
> Could you please update this to use the new formatting script: https://github.com/pika-org/pika-algorithms/blob/3e4ea7cd7f5eec8eb9c462beae0076c550aa0151/.circleci/config.yml#L104?

I don't quite understand what you're asking - are you saying that the default clang-format check should also run the string rejoin script too? - or are you asking for a change to the too/script to use something else?

**Comment by msimberg** [comment]
> I don't quite understand what you're asking - are you saying that the default clang-format check should also run the string rejoin script too? - or are you asking for a change to the tool/script to use something else?

The first. The only way we're going to _consistently_ remember to use the string-joining script is by enforcing it in CI. Otherwise it's only going to applied every now and then whenever someone remembers.

**Comment by biddisco** [comment]
> > The first. The only way we're going to consistently remember to use the string-joining script is by enforcing it in CI. Otherwise it's only going to applied every now and then whenever someone remembers.
> 
> For what I understood the string joining was a one time thing. With the fact that we now allow the line length to be 100, the strings will most likely not be broken into multiple lines no? @biddisco correct me if I'm wrong

Yes, the string join could be run every time, but it is really only needed after a change in something like clang-format rules. It might smarten up some code that has strings when added but I don't think it needs to be run always.

I'd be fine with merging 'as is' (if you'e happy) and leaving it out of CI for the time being - there is one outstanding corner case with strings that could cause a fail (and then every commit would trigger it in CI) that needs to be fixed - then it could be added painlessly to CI.

**Comment by msimberg** [comment]
Ok, thanks @aurianer and @biddisco for the comments. Let's leave it out for now then. I would prefer it to be in CI eventually, but if you're not completely happy with it yet let's leave that for another day.

**Review by msimberg — APPROVED** [review]

**Comment by msimberg** [comment]
Merging manually as CI is not running at the moment.

**Pull request merged**